gdrive: Service account support; Default credentials; Trash only conf… (#1096)
iterative/dvc.org: content/docs/user-guide/setup-google-drive-remote.md

to establish GDrive remote connections (e.g. CI/CD).

## Quick start

To start using a Google Drive remote, you only need to add it with a
[valid URL format](#url-format). Then use any DVC command that needs it (e.g.
`dvc pull`, `dvc fetch`, `dvc push`). For example:

```dvc
$ dvc remote add -d gdremote gdrive://0AIac4JZqHhKmUk9PDA/dvcstore
$ dvc push

Go to the following link in your browser:

    https://accounts.google.com/o/oauth2/auth # ... copy this link

Enter verification code: # <- enter resulting code
```

See [Authorization](#authorization) for more details.

## URL format

There's a few alternatives to construct a GDrive remote URL for different uses,
such as a folder or subfolder in root, shared folders not owned by your account,
etc. The URL is formed with a _base_, and an optional _path_ to an **existing**
folder i.e. `gdrive://<base>/path/to/folder`. The base can be one of:

1. _Folder ID_ - unique identifier for every Google Drive folder, including
   [shared folders](https://support.google.com/drive/answer/7166529) and
   [shared drives](https://support.google.com/a/users/answer/9310351) (these two
   can only be referenced by ID).

   ```dvc

>    $ dvc remote add gdfolder gdrive://0AIac4JZqHhKmUk9PDA

Reviewer: should we add `-d` here and in other places?

Reviewer 2: I intentionally only added it in the Quick start section where it's follower by `dvc push`. I don't think that it adds value in these other examples.

Reviewer: It's not obvious to be honest, if we use `-d` in all other places, let's use it everywhere?

Also, still don't see any value in these special names (one place is left).

Reviewer 2: Names updated now. I accidentally missed those, thanks.

I would rather remove -d from everywhere? The topic of default remotes seems totally separate fom gdrive to me. But then if I remove it from the Quick start example, I then need to add it to `dvc push -r myremote`... Will do that for now... ⏳

Reviewer: let's def not do `dvc push -r`, please ... better to include `-d` everywhere ... or even keep as is for now

Reviewer 2: I changed it to dvc remote add --default just in Quick start for now. That way it's more obvious and no need to explain -d or add it everywhere in this doc?

I noticed the dvc push cmd ref examples need an update and I already started another regular updates PR so I can address it there 🙂

Reviewer 2: (See #1110)

   ```

   or

   ```dvc

>    $ dvc remote add gdsubfolder \

Reviewer: I'mt not sure about using these names like gdsubfolder, etc - It might look like it is some convention. Probably it's better to keep it simple everywhere - `myremote`

Reviewer 2: OK, changed.

to establish GDrive remote connections (e.g. CI/CD).

## Quick start

To start using a Google Drive remote, you only need to add it with a
[valid URL format](#url-format). Then use any DVC command that needs it (e.g.
`dvc pull`, `dvc fetch`, `dvc push`). For example:

```dvc
$ dvc remote add -d myremote gdrive://0AIac4JZqHhKmUk9PDA/dvcstore
$ dvc push

Go to the following link in your browser:

    https://accounts.google.com/o/oauth2/auth # ... copy this link

Enter verification code: # <- enter resulting code
```

See [Authorization](#authorization) for more details.

## URL format

There's a few alternatives to construct a GDrive remote URL for different uses,
such as a folder or subfolder in root, shared folders not owned by your account,
etc. The URL is formed with a _base_, and an optional _path_ to an **existing**
folder i.e. `gdrive://<base>/path/to/folder`. The base can be one of:

1. _Folder ID_ - unique identifier for every Google Drive folder, including
   [shared folders](https://support.google.com/drive/answer/7166529) and
   [shared drives](https://support.google.com/a/users/answer/9310351)\* (these
   two can only be referenced by ID).

   ```dvc

>    $ dvc remote myremote gdfolder gdrive://0AIac4JZqHhKmUk9PDA

Reviewer: `dvc remote _add_`

Reviewer 2: Fixed, thanks. Also a couple lines down.

# Setup a Google Drive DVC Remote

In this guide we explain the existing ways to setup Google Drive
[remote storage](/doc/command-reference/remote) for your <abbr>DVC
projects</abbr>, along with the different benefits each one brings.

> Note that to start using a GDrive remote, you only need to add it with a

Reviewer: Let's do a "Quick start", "Basic config" or whatever 

It should start with:

`To start using Gdrive remote ...`

Authorization section should be just part of it

Reviewer 2: Created header and other small adjustments in 3692cf09. But let's decide about the Auth section in https://github.com/iterative/dvc.org/pull/1096#pullrequestreview-387659095 below.

[valid URL format](#url-format-to-specify-folder-location). Then use any DVC
command that needs it (e.g. `dvc pull`, `dvc fetch`, `dvc push`). For example:

```dvc

> $ dvc remote add -d gdremote gdrive://0AIac4JZqHhKmUk9PDA/dvcstore

Reviewer: we can explain below that `0AIac4JZqHhKmUk9PDA` is a folder id and check it out how to get it

Reviewer 2: Done as well.

$ dvc push
...
Go to the following link in your browser:

    https://accounts.google.com/o/oauth2/auth # ... copy this link

Enter verification code: # <- enter resulting code
```

> See the [Authorization](#authorization) section for more details on connecting

Reviewer: this separate authorization still feels strange and not needed

Reviewer 2: Well, the Authorization section is pretty long for the "Quick" start, and applies to any URL format and even with a custom GC project. I think that's why it's separate.

Reviewer 2: In fact there's a link to it from the Custom GC project section.

Reviewer 2: p.s. Also moved the Auth section back to the bottom since custom gc project and service acct sections are mentioned before quick start now, per https://github.com/iterative/dvc.org/pull/1096#pullrequestreview-387659149.

DVC with Google Drive.

> DVC uses the Google Drive API to synchronize your <abbr>DVC project</abbr> data

Reviewer: these two paragraphs can stay at the very beginning, before Quick start

Reviewer 2: Moved

with this type of remote storage, so it's subject to certain usage limits and
quotas, which by default are shared with other GDrive remote storage users. For
heavy use, it's highly recommended to
[connect using a custom Google Cloud project](#connect-using-a-custom-google-cloud-project),
which puts you in control of these limits.

Having your own GC project, it's also possible to
[use a service account](#using-service-accounts) for automating tasks that need
to establish GDrive remote connections (e.g. CI/CD).

> ## URL formats to specify folder location

Reviewer: ... to specify remote (or storage?) location?

Reviewer 2: TBH I think this should just be "URL format". We're already in the "Setup a Google Drive DVC Remote" doc (title). So it's the URL format for Google Drive DVC remotes.

I would maybe change the title to **Setup a Google Drive Remote Storage** though.

There's a few alternatives to construct a GDrive remote URL for different uses,
such as a folder or subfolder in root, shared folders not owned by your account,
etc. The URL is formed with a _base_, and an optional _path_ to an **existing**
folder i.e. `gdrive://<base>/path/to/folder`. The base can be one of:

> 1. _Folder ID_ - unique identifier for each Google Drive folder

Reviewer: probably mention shared drive, shared folder here

Reviewer 2: It is mentioned in this point, after the console samples. Not on the top because every GD folder has an ID, not just shared ones.

Reviewer 2: OK I actually went ahead and moved the note up to this paragraph.

   To obtain it, navigate to that folder in your web browser, and find it in the
   address bar. For example, for
   `https://drive.google.com/drive/folders/0AIac4JZqHhKmUk9PDA`:

   ```dvc
   $ dvc remote add gdfolder gdrive://0AIac4JZqHhKmUk9PDA

>    $ dvc remote add gdsubfolder \

Reviewer: these two commands should be split with `or`

                  gdrive://0AIac4JZqHhKmUk9PDA/Data/text
   ```

   Note that [shared folders](https://support.google.com/drive/answer/7166529)
   and [shared drives](https://support.google.com/a/users/answer/9310351) can
   only be referenced this way.

2. `root` - indicates your topmost Google Drive directory.

   ⚠️ Only suitable for personal use, as sharing a remote configured this way
   would cause DVC to try synchronizing data to/from different Google Drives for
   every user.

   ```dvc
   $ dvc remote add mygdfolder gdrive://root/dvcstore
   ```

>    > Although valid, we don't recommend using just `gdrive://root`, as it's not

Reviewer: it will be very hard to understand I believe

Reviewer 2: Changed to
   > We don't recommend using `gdrive://root` itself, as it's likely used for
   > many other reasons, and pushing data with DVC here can make it seem messy.

WDYT?

Reviewer: 👍

   > possible for more than one Google account to access a GDrive root ("My
   > Drive").

3. `appDataFolder` -

>    [special hidden folder](https://developers.google.com/drive/api/v2/appdata)

Reviewer: also suitable for personal case only

Reviewer 2: Good catch. Updated.

   (unique per user) meant to store application-specific data. This is a good
   choice to prevent accidentally deleting remote storage data from the Google
   Drive web UI.

   ```dvc
   $ dvc remote add gdappata gdrive://appDataFolder
   ```

## Authorization

On the first usage of a GDrive [remote](/doc/command-reference/remote), for
example when trying to `dvc push` for the first time after adding the remote
with a [valid URL](#url-format-to-specify-folder-location), DVC will prompt you
to visit a special Google authorization web page. There you'll need to sign into
your Google account. The
[auth process](https://developers.google.com/drive/api/v2/about-auth) will ask
you to grant DVC the necessary permissions, and produce a verification code

> needed for DVC to complete the connection. On success, this code will be cached

Reviewer: I don't think it save the code in the file

Reviewer 2: I had just changed "token" to "code" here, but left Max' wording. I see the code is not saved indeed. It's just a one-time thing. Updated text now, per https://github.com/iterative/dvc.org/pull/1096#pullrequestreview-387659574 above.

in a Git-ignored directory located in `.dvc/tmp/gdrive-user-credentials.json`.

⚠️ In order to prevent unauthorized access to your Google Drive, **do not share

> your verification code with others**. Each team member should go through this

Reviewer: ... don't share produced credentials file (not sure sharing verification code is a big deal)

Reviewer 2: Good catch. Fixed!

   ```dvc
   $ dvc remote add gdfolder gdrive://0AIac4JZqHhKmUk9PDA
   $ dvc remote add gdsubfolder \
                  gdrive://0AIac4JZqHhKmUk9PDA/Data/text
   ```

   Note that [shared folders](https://support.google.com/drive/answer/7166529)
   and [shared drives](https://support.google.com/a/users/answer/9310351) can
   only be referenced this way.

2. `root` - indicates your topmost Google Drive directory.

   ⚠️ Only suitable for personal use, as sharing a remote configured this way
   would cause DVC to try synchronizing data to/from different Google Drives for
   every user.

   ```dvc
   $ dvc remote add mygdfolder gdrive://root/dvcstore
   ```

   > Although valid, we don't recommend using just `gdrive://root`, as the
   > account's Google Drive home ("My Drive") could fill with cache directories,
   > complicating managing non-DVC files and folders.

3. `appDataFolder` -
   [special hidden folder](https://developers.google.com/drive/api/v2/appdata)
   (unique per user) meant to store application-specific data. This is a good
   choice to prevent accidentally deleting remote storage data from the Google
   Drive web UI.

   ⚠️ Only suitable for personal use.

   ```dvc

>    $ dvc remote add gdappata gdrive://appDataFolder

Reviewer 2: Should it be lower case, something like `gdrive://appdata` instead, maybe? I don't think we use cammelCase much in these kind of literals inside DVC in general. Not "pythonic", I think. Obviously not a huge deal.

@shcheklein @maxhora

Reviewer: `appDataFolder` is a predefined Google Drive specific name, it's not something we control. Same as `root`.

Reviewer 2: Yes, makes sense. But the remote URL is an internal thing, they could be mapped in the code to what Google expects.

Reviewer: this is a good point. If we support root and appData I would still keep names as is - to be more explicit. It might be though for simplicity we can just decide to support only URLs like: 

`gdrive://<folder-id>` in the future to make less error prone.

at least it feels like we should recommend this approach as much as possible.

Reviewer 2: Agree.

> we should recommend this approach

Added "recommended" in folder id, and both other ways have warnings about being only good for personal use, so I think it's clear now.

# Setup a Google Drive DVC Remote

> In this guide we explain the existing alternatives to setup Google Drive

Reviewer: minor: alternatives -> ways (a bit easier to read, since other wise it's easy to misread as an alternative to Google Drive)

Reviewer: we don't explain Basic setup ... let's do a quick start even before URL, or even overview that includes the basic setup (and even authorization) - like very simple a quick? May be very similar to what we have in `dvc remote add` now. And a few links to URLs and "Using your won project" sections?

Reviewer 2: Using "ways".
And I did already include a paragraph starting in "Note that to start using a GDrive remote, you only need..." but OK, copied the basic example from `remote add` now.
64e85917

[remote storage](/doc/command-reference/remote) for your <abbr>DVC
project</abbr>, along with the different benefits each one brings.

> Note that to start using a GDrive remote, you only need to add it with a valid

Reviewer: see above, I think we need a quick setup or basic setup or something

this whole doc should be self-contained, so should include `dvc remote add` in the very beginning.

Reviewer 2: I just added a basic example under this paragraph now, would that be enough? Our user guides aren't typically super self-contained or fully reproducible like tutorials though, we do link to the cmd refs.

Reviewer: yes, it's enough .. I was thinking about being more explicit and making a title like "Quick start" or "Basic setup"

```

2. Google Drive _Folder ID_

To obtain the unique _Folder ID_ in question, navigate into that folder in your
web browser, and find it in the address bar. For example, for
`https://drive.google.com/drive/u/0/folders/0AIac4JZqHhKmUk9PDA`:

```dvc
$ dvc remote add mygdfolder gdrive://0AIac4JZqHhKmUk9PDA
$ dvc remote add mygdsubfolder \
                 gdrive://0AIac4JZqHhKmUk9PDA/sub/path
```

Note that
[Shared drives](https://developers.google.com/drive/api/v2/about-shareddrives)
can only be referenced by _Folder ID_.

3. `appDataFolder` -
   [special hidden folder](https://developers.google.com/drive/api/v2/appdata)
   that can be used to store application-specific data.

```dvc
$ dvc remote add mygdappata gdrive://appDataFolder
```

## Configure custom Google Cloud project & app (recommended)

Optionally, follow this guide to setup your own Google Cloud project and
generate OAuth credentials for your <abbr>DVC projects<abbr> to connect to
Google Drive. Doing so avoids sharing our default GCP project/app with every
other casual user of GDrive remotes out there. This gives you full control over
Google API usage and rate limits, and ensures optimal performance to access your

> Google Drive with DVC. This is highly recommended for heavy and enterprise use.

Reviewer: > This is highly recommended for heavy and enterprise use.

Not sure about enterprise. Let's just keep heavy, frequent, or something? Should be enough?

The whole paragraph is too technical and complicated. Can we explain it in some simple way?

Reviewer: Simple cases:

* limits (Retying something ... message should be visible in logs), possibility to ask Google to extend limits independently - they answer quickly
* use service accounts (e.g. some automation, CI/CD, etc)

*

Reviewer 2: Updated with a bullet list, please check.

> > Please jump to [Authorization](#authorization) to skip this setup.

Reviewer: why skip? why jump? not clear

Reviewer 2: Because it's optional but a pretty long section, nice to have a link to the next step?

Reviewer: so, let's put Authorization before all optional sections?

Reviewer 2: Yeah you're right. Moved under **URL format** section.

Reviewer: it's still here?

Reviewer 2: The note has been removed. The sections were moved around a few times to match the order in the intro and **Quick start**.

# Setup a Google Drive DVC Remote

In this guide we explain the existing ways to setup Google Drive
[remote storage](/doc/command-reference/remote) for your <abbr>DVC
projects</abbr>, along with the different benefits each one brings.

Note that to start using a GDrive remote, you only need to add it with a valid

> URL (see next section), and use any DVC command that requires it (see

Reviewer: see next section -> see below how to specify it for shared drive, shared folders, etc ? be a bit more explicit why would I need to take a look there

Reviewer 2: I agree with giving more motivation on URL formation, but I added it in the URL format section and just linked from here instead. See 9d9ebec4. Good?

Reviewer: kk, it's fine, but see other comment about the basic case ... it feels that it's wrong that we use `root` and if we switch to something like `gdrive://FOLDER_ID/path` section that explains URLs becomes even more important. And explaining limitations for different URLs becomes very important (e.g. for `grdive://root` it won't be easy to share, and can be even dangerous since `gdrive://root` will be resolved fine and multiple people will push data to two different GD accounts)

Reviewer 2: Yes, good points. Will address along with what I summarized in https://github.com/iterative/dvc.org/pull/1096#pullrequestreview-386926510.

`dvc pull`, `dvc fetch`, `dvc push`). For example:

```dvc
$ dvc remote add -d myremote gdrive://root/dvc/dvcstore
$ dvc add file.dat
$ dvc push -r myremote
...
Go to the following link in your browser:

    https://accounts.google.com/o/oauth2/auth # ... copy this link

Enter verification code: # <- enter resulting token
```

See the [Authorization](#authorization) section for more details.

> ## URL format for Google Drive remotes

Reviewer: do we need to mention that it's Google Drive remotes? We are already in the doc ..

would be better to be explicit here as well - Specifying location ?

Reviewer 2: Good point, updated in 34f8b410.

A GDrive remote URLs is needed for the `dvc remote add` command. It can be
constructed with a _base_, and an **optional** _path_ to a folder i.e.
`gdrive://<base>/path/to/folder`. The base can be one of:

1. `root` - indicates your topmost Google Drive directory.

```dvc

> $ dvc remote add -d mygdroot gdrive://root

Reviewer: let's not use empty `//root` as an example - we should probably raise an exception if it's used

Reviewer 2: Shouldn't it be up to the user? If it's not a good practice then I'll use a different example ⏳ but not sure about the exception. Maybe a warning. Anyway, that's a core repo issue 🙂

Reviewer 2: Removed line, and opened dvc/issues/3586

Reviewer: thanks! a good one. Made me think once again about basic scenario

$ dvc remote add mygdfolder gdrive://root/path/to/folder
```

2. Google Drive _Folder ID_

To obtain the unique _Folder ID_ in question, navigate into that folder in your
web browser, and find it in the address bar. For example, for
`https://drive.google.com/drive/u/0/folders/0AIac4JZqHhKmUk9PDA`:

```dvc
$ dvc remote add mygdfolder gdrive://0AIac4JZqHhKmUk9PDA
$ dvc remote add mygdsubfolder \
                 gdrive://0AIac4JZqHhKmUk9PDA/sub/path
```

Note that [shared folders](https://support.google.com/drive/answer/7166529) and
[shared drives](https://support.google.com/a/users/answer/9310351) can only be
referenced by _Folder ID_.

3. `appDataFolder` -
   [special hidden folder](https://developers.google.com/drive/api/v2/appdata)
   (unique per user) meant to store application-specific data. This is a good
   choice to prevent accidentally deleting remote storage data from the Google
   Drive web UI.

```dvc
$ dvc remote add mygdappata gdrive://appDataFolder
```

> ## Configure custom Google Cloud project & app (recommended)

Reviewer: we should mention it briefly at least in the intro? 

also, I still don't like the title.

Ideally it should go from the task people would probably trying to solve with this ... not sure how to phrase this in a simple way

How about:

Using custom project
Use custom credentials
Connect with custom project
etc

(recommended) - better to highlight it here, or in the note on the top that mentions this section and provides and why would I need this

Reviewer 2: Alright. Title updated and note added in intro to this section, as well as to the service account section.

Reviewer 2: p.s. the intro is kinda long now...

Reviewer: yep, will check it again

Optionally, follow this guide to setup your own Google Cloud project and
generate OAuth credentials for your <abbr>DVC projects<abbr> to connect to
Google Drive. We highly recommend this for heavy and advanced use because:

- you control your Google API usage and rate limits, being able to request an

>   increase to Google if needed.

Reviewer: to request .. to? do we need a second to? should it be just "request Google to increase"?

Follow this guide to setup Google Drive as your <abbr>DVC project</abbr>'s
[remote storage](/doc/command-reference/remote).

> ## Configure a Google Cloud project (Highly recommended optional step)

Reviewer 2: ```suggestion
## Configure a Google Cloud project (highly recommended)
```

Author: I'm not sure what phrase fits the desired balance here - it seems we highly recommend to configure their own GCP project, but we should also highlight that the step is optional and user can save a lot time by skipping it.

Reviewer 2: Recommended implies optional 🙂

Reviewer 2: Also I added the term "optionally" in my next suggestion. Thanks

We highly recommend to follow this guide and complete setup of your own Google

> Cloud project and

Reviewer 2: ```suggestion
Optionally but highly recommended, follow this guide to setup your own Google Cloud project and
```

Reviewer 2: We should also probably expand on why this is beneficial to the user in another sentence in this same paragraph. E.g. he will not share rate limits with others, maybe more control/visibility for debugging, Idk, etc.

Reviewer: yes, we def need to add motivation, mention that sharing potentially could lead for slower performance, give hints how to check if users hit limits, etc

Reviewer 2: > give hints how to check if users hit limits

I'm not sure how to do this.

Could you guys help me summarize the benefits of crating your own GCP project?

I added some general info for now but it's probably not enough, please check it out.

Reviewer 2: p.s. you know what I should just try all the gdrive config options again to help in QA which is probably needed anyway, I think that will allow me to review this and the whole doc anyway ⏳

Reviewer 2: OK, added some precisions, including about this now.

Follow this guide to setup Google Drive as your <abbr>DVC project</abbr>'s
[remote storage](/doc/command-reference/remote).

## Configure a Google Cloud project (highly recommended)

Optionally, follow this guide to setup your own Google Cloud project and

> [OAuth](https://developers.google.com/identity/protocols/OAuth2) credentials.

Reviewer 2: OK. I'm not seeing how this is optional. You need a GC project to use either client creds (id/secret) or a service account. How can this be skipped @maxhora ? I.e. how do you use the default DVC project? Thanks

Author: @jorgeorpinel the change to use client id and secret from the default DVC project was made in the last my commit there https://github.com/iterative/dvc/commit/8a4827017ea171ef0dab49cb79785c997075bab6#diff-21bfd10d01c6715fc8634644730fa720R102
If user haven't specified client_id and client_secret in DVC config the default pair will be used by DVC (user will not need to create their own GCP project in this case).

Reviewer: Ping me on Slack, if it's still not clear, @jorgeorpinel

Reviewer 2: Got it. It's just not clear that you don't have to provide anything other than the gdrive URL, and then the first `dvc push` will give you a link to authorize the DVC app:

![image](https://user-images.githubusercontent.com/1477535/77976738-fea60c80-72ba-11ea-8d67-77133f02cb6f.png)

But I got it now I think. I'll clarify all this, thanks.

Reviewer: Yes! Google drive is in fact the most simple of all remotes to start - and that's what we should clearly communicate. With GD you can start with just an URL. Then if you hit limits (you can check logs and see the message), you should create your own project. Service account and stuff - advanced features.

Also, might potentially check rclone's docs - how they document their shared credentials and how they motivate user on creating their own projects.

Reviewer 2: OK I've mostly addressed this except didn't mention this progression about first just use URL, if you hit limits create a GC project, etc. Didn't check rclone docs yet either but will do ⏳

Reviewer 2: I think this is addressed now. I did also check out https://rclone.org/drive/ but didn't find much inspiration anyway.

[authorization](#authorization) URL DVC will later prompt to visit in order to
connect to the Google Drive.

> Note that the Google Drive API has usage limits/quotas per _project_ client
> (which you can review in the
> [OAuth consent screen](https://console.developers.google.com/apis/credentials/consent)).
> Please keep this in mind when sharing them, or you may
> [exceed the limits](https://developers.google.com/drive/api/v2/handle-errors?hl=ro#resolve_a_403_error_usage_limit_exceeded).

### Access Google Drive remote with OAuth credentials

Use the `dvc remote modify` command to set the credentials for each `gdrive://`
remote, for example:

```dvc
$ dvc remote add -d mygdfolder gdrive://root/path/to/folder
$ dvc remote modify mygdfolder gdrive_client_id <client ID>
$ dvc remote modify mygdfolder gdrive_client_secret <client secret>
```

> ### Access Google Drive remote with a service account

Reviewer: Simplify title - `Using service accounts`

Make it a separate section, put a link to it in the `Creating you app` and back to `Creating your app` as a prerequisite.

Reviewer 2: OK, done.

$ dvc remote modify mygdfolder gdrive_client_secret <client secret>
```

> ## Configure the GDrive remote with a service account

Reviewer 2: ```suggestion
## Configure the Google Drive remote with a service account
```
Same change for the previous header.

Reviewer 2: Also, what's the difference in the previous header? Should it be something like **Configure the Google Drive remote with user credentials** to make the difference clear at first glance?

Author: DVC should be configured with different params in config to auth user account (client id and secret should be in the config) and to auth service account (service account email address and .p12 local key file path params in the config). Renaming previous section header sounds great!

Reviewer 2: Thanks for the info.

A
[service account](https://cloud.google.com/docs/authentication/getting-started#creating_a_service_account)

> can be used to configure the GDrive remote.

Reviewer 2: What's a service account? What are the benefits?

Author: `dvc remote modify` already mentions `service account` for Google Cloud. It is the same thing for Google Drive as well.
GCP project can have list of service accounts. Each service account has its own email address and some granted access rights, etc.
Benefit to use service account is to remove dependency on some user's account. Also it is recommended way to do some automated actions, like running tests on Travis.

Reviewer 2: Got it. I also found https://cloud.google.com/iam/docs/understanding-service-accounts. I'll review the GC section and maybe copy some info into this one.

[authorization](#authorization) URL DVC will later prompt to visit in order to
connect to the Google Drive.

> Note that the Google Drive API has usage limits/quotas per _project_ client
> (which you can review in the
> [OAuth consent screen](https://console.developers.google.com/apis/credentials/consent)).
> Please keep this in mind when sharing them, or you may
> [exceed the limits](https://developers.google.com/drive/api/v2/handle-errors?hl=ro#resolve_a_403_error_usage_limit_exceeded).

> ### Enabling your GC project with OAuth credentials

Reviewer: I don't quite understand this title ... why do we enable GC project?

Reviewer: make it - `Configuring DVC to use you custom credentials`?

Reviewer 2: Yeah. How about we just remove the header actually? I feel like it's simply part of the custom gc project section anyway. See f6b1aebc.

Reviewer 2: p.s. the intro is kinda long now...

⚠️ It should be safe to share **client ID** and **client secret** among your
team. These credentials are only used to generate the
[authorization](#authorization) URL you'll need to visit later in order to
connect to the Google Drive.

Finally, use the `dvc remote modify` command to set the credentials (for each
GDrive remote), for example:

```dvc

> $ dvc remote add gdfolder gdrive://0AIac4JZqHhKmUk9PDA

Reviewer: probably we should not mention this command here - better to mention basic workflow and specify additional commands only

Reviewer 2: Removed.

[authorization](#authorization) URL DVC will later prompt to visit in order to
connect to the Google Drive.

Finally, use the `dvc remote modify` command to set the credentials (for each
GDrive remote), for example:

```dvc
$ dvc remote add gdfolder gdrive://0AIac4JZqHhKmUk9PDA
$ dvc remote modify gdfolder gdrive_client_id <client ID>
$ dvc remote modify gdfolder gdrive_client_secret <client secret>
```

> Note that Google Drive API usage limits/quotas apply per _project_ client and
> can be reviewed in the
> [OAuth consent screen](https://console.developers.google.com/apis/credentials/consent).
> Please keep this in mind when sharing them, or you may
> [exceed the limits](https://developers.google.com/drive/api/v2/handle-errors?hl=ro#resolve_a_403_error_usage_limit_exceeded).

## Using service accounts

A [service account](https://cloud.google.com/iam/docs/service-accounts) is a
Google account associated with your GCP project, and not a specific user. They
are intended for scenarios where your code needs to access data on its own, e.g.
running inside a Compute Engine, automatic CI/CD, etc. No interactive user OAuth
authentication is needed.

> This requires having your own

> > [GC project](/doc/user-guide/setup-google-drive-remote#connect-using-a-custom-google-cloud-project)

Reviewer: link is probably broken - please check other places

Reviewer 2: I've been keeping them current.

> as explained above.

1. To
   [create a service account](https://cloud.google.com/docs/authentication/getting-started#creating_a_service_account),
   navigate to **IAM & Admin** in the left sidebar, and select **Service
   Accounts**. Click **+ CREATE SERVICE ACCOUNT**, on the next screen, enter
   **Service account name** e.g. "My DVC project", and click **Create**. Select
   **Continue** at the next **Service account permissions** page, click at **+
   CREATE KEY**, select **P12** and **Create**. Save generated `.p12` key file
   at your local disk.

> 1. Copy a downloaded `.p12` file to your DVC project root directory.

Reviewer: this is a bad idea I think, this file must never be put into Git ... so probably should stay in some safe local location

Reviewer 2: Ok, updated. Also added this note:

⚠️ Be careful about sharing the key file with others.

Reviewer: 👍

are intended for scenarios where your code needs to access data on its own, e.g.
running inside a Compute Engine, automatic CI/CD, etc. No interactive user OAuth
authentication is needed.

```dvc
$ dvc remote add mygdappata gdrive://appDataFolder
```
> This requires having your own
> [GC project](/doc/user-guide/setup-google-drive-remote#connect-using-a-custom-google-cloud-project)
> as explained above.

## Configure the GDrive remote
1. To
   [create a service account](https://cloud.google.com/docs/authentication/getting-started#creating_a_service_account),
   navigate to **IAM & Admin** in the left sidebar, and select **Service
   Accounts**. Click **+ CREATE SERVICE ACCOUNT**, on the next screen, enter
   **Service account name** e.g. "My DVC project", and click **Create**. Select
   **Continue** at the next **Service account permissions** page, click at **+
   CREATE KEY**, select **P12** and **Create**. Save generated `.p12` key file
   at your local disk.

Use the `dvc remote modify` command to set the credentials for each `gdrive://`
remote, for example:
1. Copy a downloaded `.p12` file to your DVC project root directory.

```dvc

> $ dvc remote add -d mygdfolder gdrive://root/path/to/folder

Reviewer: it looks like we are missing `3. Setup DVC to use ...`

Reviewer 2: I'm not sure on the location of this note since it's left in a removed line and the line count has changed a lot. But if this was about the Service Acct section, I think it's no longer needed.

> $ dvc remote modify gdremote gdrive_use_service_account True

Reviewer: True -> true?

Reviewer 2: Yes, fixed.
